pkg/engine/supervisor.go

```diff
@@ -70,13 +70,42 @@ func (e *Engine) worker(n int) {
 
 			ch := make(chan int)
 			e.addSignal(tsk.ID, ch)
-
 			go func() {
 				select {
 				case <-ch:
 					e.deleteSignal(tsk.ID)
 					cancel()
 				case <-ctx.Done():
+					// Timeout reached,
+					// 1. Remove timeouted tasks from the queue -- It is going to be removed anyway, since we are popping from the queue.
+					// 2. Stop timeouted task execution (e.g. stopping the docker container)
+					e.deleteSignal(tsk.ID)
+					// 3.  Update the state of the task in the database
+					logging.S().Infow("Timeout reached for task: ", "task_id", tsk.ID)
+					// get the current state of the task from DB.
+					tsk, err := e.store.Get(tsk.ID)
+					if err != nil {
+						return
+					}
+					// if task is still in processing state, mark it as canceled in DB
+					if tsk.State().State == task.StateProcessing {
+						canceledState := task.DatedState{
+							Created: time.Now().UTC(),
+							State:   task.StateCanceled,
+						}
+						tsk.States = append(tsk.States, canceledState)
+						logging.S().Infow("Task state: ", "state", tsk.State().State)
+
+						err = e.store.PersistProcessing(tsk)
+						if err != nil {
+							logging.S().Errorw("could not persist task", "err", err)
+							return
+						}
+
+						if tsk.IsCanceled() {
+							logging.S().Infow("Task reached timeout", "task_id", tsk.ID)
+						}
+					}
```

Comment:
Where do we stop the containers?

Reply:
As I understood, when the signal is deleted  from	`signals   map[string]chan int`  runner will stop all containers linked with that task.ID, correct me if I am wrong?

Reply:
I guess the same way that Kill also works

```
// Kill closes the signal channel for a given task, which signals to the runner to stop it
func (e *Engine) Kill(id string) error {
	e.signalsLk.RLock()
	if ch, ok := e.signals[id]; ok {
		close(ch)
	}
	e.signalsLk.RUnlock()

	return nil
}
```

Reply:
Although I see that there is method over client that can stop container, I can not get to call that method from the right place where context is canceled, since 'engine' is not having that client as the dependency 

```
func (cli *Client) ContainerStop(ctx context.Context, containerID string, timeout *time.Duration) error {
   query := url.Values{}
   if timeout != nil {
      query.Set("t", timetypes.DurationToSecondsString(*timeout))
   }
   resp, err := cli.post(ctx, "/containers/"+containerID+"/stop", query, nil, nil)
   ensureReaderClosed(resp)
   return err
}
```

Reply:
I can't find that command in the codebase, https://github.com/testground/testground/search?q=ContainerStop,
It does look like an API, that's an entry point that the daemon exposes to clients, we don't want to use this internally.

Could you share links to the code? We might not be working on this part of the codebase when we review, and it's taking us time to move around the codebase even though you already did it.

You can also share permalink: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet

Like this one: https://github.com/testground/testground/blob/63a8fce8cf08d576e745186dc322312ccf3af97c/pkg/engine/engine.go#L419-L427

The code you shared closes the channel, but are we sure closing this channel will kill the containers/pods/process in docker/k8s/exec ?

Reply:
I am providing you screenshots of docker containers running step by step

1. testground daemon
<img width="1109" alt="Screenshot 2022-09-22 at 14 11 24" src="https://user-images.githubusercontent.com/47892729/191744097-76e177fb-68d6-4120-b1b5-bb85155cf122.png">

2. running test case` testground run single --plan integrations --testcase issue-1432-task-timeout  --builder docker:go --runner local:docker --instances 1`
<img width="1178" alt="Screenshot 2022-09-22 at 14 12 07" src="https://user-images.githubusercontent.com/47892729/191744259-b9a98225-f72e-44d1-b131-2018046594b4.png">
<img width="1174" alt="Screenshot 2022-09-22 at 14 12 53" src="https://user-images.githubusercontent.com/47892729/191744340-f1c8e2cf-0e76-46ec-9dae-1d38d2674449.png">

3. Task timeouted after 1 minute



<img width="1184" alt="Screenshot 2022-09-22 at 14 13 18" src="https://user-images.githubusercontent.com/47892729/191745166-a38ff4ba-2b9f-4495-9069-aeefad5fc321.png">

<img width="1188" alt="Screenshot 2022-09-22 at 14 13 32" src="https://user-images.githubusercontent.com/47892729/191745257-56c9ae22-8096-4111-8316-083d87ba3a3b.png">
<img width="1108" alt="Screenshot 2022-09-22 at 14 13 43" src="https://user-images.githubusercontent.com/47892729/191745299-ffcbc573-0a6d-4c89-b63b-50037e34b342.png">

Reply:
Parent context is being used to create new context with cancel
https://github.com/testground/testground/blob/bb4451ee7bf0f1542278eabbdd5559667ca223a4/pkg/runner/local_docker.go#L486
that is used for starting containers:
`err := cli.ContainerStart(startGroupCtx, c.containerID, types.ContainerStartOptions{})`

this defered function is stopping running containers:
https://github.com/testground/testground/blob/bb4451ee7bf0f1542278eabbdd5559667ca223a4/pkg/runner/local_docker.go#L458

@laurentsenta

Reply:
I'll need more time to review this, could you fix the other requests? Thanks in advance, I'll try to have a second pass tomorrow.